src/server/worker/datum/create_stream.go

```diff
@@ -0,0 +1,93 @@
+package datum
+
+import (
+	"context"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/client"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/renew"
+	"github.com/pachyderm/pachyderm/v2/src/internal/task"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+	"github.com/pachyderm/pachyderm/v2/src/pps"
+)
+
+const (
+	// The target number of files per shard. A smaller value returns datums to
+	// the client faster. A larger value reduces the overhead of creating datums.
+	ShardNumFiles = 10000
+)
+
+type createDatumStream struct {
+	ctx               context.Context
+	c                 pfs.APIClient
+	taskDoer          task.Doer
+	input             *pps.Input
+	requestDatumsChan chan bool
+	fsidChan          chan string
+	errChan           chan error
+}
+
+func (cds *createDatumStream) create(input *pps.Input, fsidCh chan string) {
+	switch {
+	case input.Pfs != nil:
+		cds.createPFS(input.Pfs, fsidCh)
+	case input.Union != nil:
+		cds.errChan <- errors.New("union input type unimplemented")
+	case input.Cross != nil:
+		cds.errChan <- errors.New("cross input type unimplemented")
+	case input.Join != nil:
+		cds.errChan <- errors.New("join input type unimplemented")
+	case input.Group != nil:
+		cds.errChan <- errors.New("group input type unimplemented")
+	case input.Cron != nil:
+		cds.errChan <- errors.New("can't create datums for cron input type")
+	default:
+		cds.errChan <- errors.Errorf("unrecognized input type: %v", input)
+	}
+}
+
+func (cds *createDatumStream) createPFS(input *pps.PFSInput, fsidChan chan string) {
+	defer close(fsidChan)
+	authToken := getAuthToken(cds.ctx)
```

Comment:
This helper was kind of added as a hack to fix a bug, but isn't a great design in general.  It is not aware whether auth is on or off, so when auth is off and there is no credential in the context (which is correct), it logs an error.  Since you're calling this from an RPC, you can just ask if auth is on, and if it is, require that the context have credentials.